modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java

```diff
@@ -0,0 +1,221 @@
+
+        state = STATE.QUEUEING_DATA;
+        validator.accept((HttpMessage) httpObject, new ActionListener<>() {
+            @Override
+            public void onResponse(Void unused) {
+                // Always use "Submit" to prevent reentrancy concerns if we are still on event loop
+                ctx.channel().eventLoop().submit(() -> validationSuccess(ctx));
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                final HeaderValidationException toForward;
+                if (e instanceof HeaderValidationException) {
+                    toForward = (HeaderValidationException) e;
+                } else if (e instanceof ElasticsearchException) {
+                    toForward = new HeaderValidationException((ElasticsearchException) e, true);
+                } else {
+                    toForward = new HeaderValidationException(new ElasticsearchException(e), true);
+                }
+                // Always use "Submit" to prevent reentrancy concerns if we are still on event loop
+                ctx.channel().eventLoop().submit(() -> validationFailure(ctx, toForward));
+            }
+        });
+    }
+
+    private void validationSuccess(ChannelHandlerContext ctx) {
+        assert ctx.channel().eventLoop().inEventLoop();
+        state = STATE.HANDLING_QUEUED_DATA;
+
+        int pendingMessages = pending.size();
+
+        boolean fullRequestForwarded = false;
+        HttpObject object;
+        while ((object = pending.poll()) != null) {
+            ctx.fireChannelRead(object);
+            ReferenceCountUtil.release(object);
+            if (object instanceof LastHttpContent) {
+                fullRequestForwarded = true;
+                break;
+            }
+        }
+        if (pending.size() <= 4 && pendingMessages > 32) {
+            // Prevent the ArrayDeque from becoming forever large due to a single large message.
+            ArrayDeque<HttpObject> old = pending;
+            pending = new ArrayDeque<>(4);
+            pending.addAll(old);
+        }
+
+        if (fullRequestForwarded) {
+            state = STATE.WAITING_TO_START;
+            if (pending.isEmpty() == false) {
+                requestStart(ctx);
+            }
+        } else {
+            state = STATE.FORWARDING_DATA;
+        }
+
+        ctx.channel().config().setAutoRead(shouldRead());
+    }
+
+    private void validationFailure(ChannelHandlerContext ctx, HeaderValidationException e) {
+        assert ctx.channel().eventLoop().inEventLoop();
+        state = e.shouldCloseChannel() ? STATE.DROPPING_DATA_PERMANENTLY : STATE.HANDLING_QUEUED_DATA;
+        HttpMessage messageToForward = (HttpMessage) pending.remove();
+        boolean fullRequestConsumed;
+        if (messageToForward instanceof LastHttpContent toRelease) {
+            messageToForward = (HttpMessage) toRelease.replace(Unpooled.EMPTY_BUFFER);
+            toRelease.release();
+            fullRequestConsumed = true;
```

Comment:
I don't understand the reason for this, can you please explain why we need to clear the buffer only if the first message is also the last one? What's special about a http request that only has a `LastHttpContent` message.

Reply:
If the first message (headers) is also the last "content", it is possible that actually byte buffer content is included in the message. We clear that here before sending the exception down the pipeline.